bot/Engine/BotController.py

```diff
@@ -0,0 +1,385 @@
+        if order is None:
+            position_id = str(uuid4())
+        elif order is not None:
+            position_id = order.position_id
+
+        new_order_model = Order(
+                                id = str(uuid4()),
+                                position_id = position_id,
+                                bot_id = self.bot.id,
+                                symbol = symbol,
+                                price = order_params['price'],
+                                take_profit_price = order_params['take_profit_price'],
+                                order_type = order_params['order_type'],
+                                stop_price = order_params['stop_price'],
+                                original_quantity = order_params['quantity'],
+                                side = order_params['side'],
+                                is_entry =  order_params['is_entry'],
+                                is_test = self.test_mode,
+                                )
+        return new_order_model
+
+    def syncModels(self, pair, order, new_order_model):
+        """ Sync pairs and orders to new status """
+        if order is not None:        
+            new_order_model.matched_order_id = order.id
+            order.is_closed = True
+            order.matched_order_id = new_order_model.id
+            pair.active = False
+            pair.current_order_id = new_order_model.id
+
+        if order is None:
+            self.bot.current_balance = self.bot.current_balance - Decimal(self.bot.starting_balance) * Decimal(self.bot.entry_settings.initial_entry_allocation) / 100
+            pair.active = False
+            pair.current_order_id = new_order_model.id    
+
+    def computeQuantity(self, order):
+        """ compute exit quantity so that also partially filled orders can be handled."""
+        if order.side == 'BUY':
+            exit_quantity = order.executed_quantity
+        elif order.side == 'SELL':
+            exit_quantity = Decimal(order.original_quantity) - Decimal(order.executed_quantity)
+        return exit_quantity
+    
+    def simulateOrderInfo(self, order):
+        """ Used when BotController is in test mode. Simulates order info returned by exchange."""
+        order_status = dict()
+        candlestick_data = self.exchange.getSymbolKlines(order.symbol, self.kline_interval, self.strategy.minimum_period)
+        current_price = candlestick_data.iloc[-1]['close']
+        if order.side == 'BUY':
+            # Market Entry Order
+            if self.bot.entry_settings.signal_distance == 0:
+                order_status['status'] = self.exchange.ORDER_STATUS_FILLED
+                order_status['side'] = order.side
+                order_status['executedQty'] = order.original_quantity
+            # Limit Entry Order
+            elif (Decimal(100) + Decimal(self.bot.entry_settings.signal_distance))/Decimal(100) * Decimal(order.entry_price) <= current_price:
```

Comment:
A bit confused by this line - what is entry_price? Isn't recalculating price + signal_distance not necessary at this point (isn't that value already stored in order)?

Reply:
At this line we want to know if the limit entry order was filled. So I agree with u that we do not need to recalculate price + signal distance, this is already stored in price or entry_price.

**entry_price:**

For entry orders entry_price and price is the same thing. But for exit orders entry_price is the price we entered a trade with. This we need to be able to check if stop-loss or take-profit conditions are met as we otherwise have no way of knowing what price we entered the trade at.

Reply:
Ok got it. Previously we used the matched_order_id for that, but I suppose entry price simplified things from a coding perspective.

Reply:
Right, also it makes the code more efficient I think. You don't have to backtrack orders by matched_order_id to the original buy order, which can save some extra queries.